Data explorer: Add support for histograms and frequency tables

dfalbel opened the pull request:
This implements the OpenRPC protocol for support computing histograms and frequency tables. This is the backend for computing sparklines in the data explorer. 

Addresses https://github.com/posit-dev/positron/issues/4061

TODO:

- [x] Add support for frequency tables (consider factors, and character vectors)
- [x] Add tests for quantile computations.
- [x] Add implementation/tests for the Sturges algorithm
- [x] Plumb with the infra to produce results on the filtered data (if any filters are applied)

lionel- commented on crates/ark/src/modules/positron/r_data_explorer.R:
```diff
@@ -337,3 +333,113 @@ write_html <- function(x, include_header) {
+    range <- max_value - min_value
+
+    if (inherits(x, "POSIXct") && as.integer(range) < num_bins) {
+     num_bins <- as.integer(range) + 1L
```
```suggestion
      num_bins <- as.integer(range) + 1L
```

lionel- commented on crates/ark/src/modules/positron/r_data_explorer.R:
```diff
@@ -337,3 +333,113 @@ write_html <- function(x, include_header) {
+  # We force something considering the integer representation.
+  if (inherits(x, "POSIXct")) {
+    if (method == "sturges") {
+        num_bins <- grDevices::nclass.Sturges(x)
```
```suggestion
      num_bins <- grDevices::nclass.Sturges(x)
```

lionel- commented on crates/ark/src/modules/positron/r_data_explorer.R:
```diff
@@ -337,3 +333,113 @@ write_html <- function(x, include_header) {
+    # so we generate our own bin_edges for them.
+    min_value <- min(x)
+    max_value <- max(x)
+    range <- max_value - min_value
```
width rather than range?

lionel- commented on crates/ark/src/modules/positron/r_data_explorer.R:
```diff
@@ -337,3 +333,113 @@ write_html <- function(x, include_header) {
+    # > In the last three cases the number is a suggestion only; as the breakpoints will
+    # > be set to `pretty` values.
+    breaks <- num_bins
+    stopifnot(is.integer(breaks) && length(breaks) == 1)
```
```suggestion
    stopifnot(is.integer(breaks), length(breaks) == 1)
```

lionel- commented on crates/ark/src/modules/positron/r_data_explorer.R:
```diff
@@ -337,3 +333,113 @@ write_html <- function(x, include_header) {
+
+    if (length(x) == 0) {
+        return(list(
+            values = c(),
+            counts = c(),
```
It seems clearer to explicitly use the `NULL` value? Esp. important to be clear that we return nullable values since these go through interop boundaries.

```suggestion
            values = NULL,
            counts = NULL,
```

lionel- commented on crates/ark/src/data_explorer/histogram.rs:
```diff
@@ -0,0 +1,579 @@
+    }
+
+    let num_bins: RObject = match params.num_bins {
+        Some(v) => (v as i32).into(),
```
Should we do these conversions with a tool that checks for boundaries? Not that it matters here but it seems like it'd be safer to get into a habit of checking bounds before truncating.

lionel- commented on crates/ark/src/data_explorer/histogram.rs:
```diff
@@ -0,0 +1,579 @@
+    where
+        RObject: From<T>,
+    {
+        let column = r_parse_eval0(code, R_ENVS.global).unwrap();
```
Might be worth adding `harp::parse_eval_global()`

lionel- replied:
Done in https://github.com/posit-dev/ark/pull/480

By the way, if you don't need the global environments (to assign into or do search path lookups), better evaluate in the base env. This way readers don't have to wonder what the global env is used for and this eliminates a range of possible side effects that could contaminate other tests. I've also added `harp::parse_eval_base()` for this.

lionel- commented on crates/ark/src/data_explorer/histogram.rs:
```diff
@@ -0,0 +1,579 @@
+    }
+
+    #[test]
+    fn test_date_histogram() {
```
Nice test util and great set of tests!

lionel- reviewed (approved):
Looks great!

DavisVaughan commented on crates/harp/src/object.rs:
```diff
@@ -603,6 +604,18 @@ impl From<&Vec<i64>> for RObject {
+        unsafe {
+            let vector = RObject::from(Rf_allocVector(REALSXP, values.len() as isize));
+            for idx in 0..values.len() {
+                SET_REAL_ELT(vector.sexp, idx as isize, values[idx] as c_double);
```
But I guess then we'd have to move from `from(values: &Vec<f64>)` to `try_from(values: &Vec<f64>)` so maybe we shouldn't worry about that right now.

(i do think maybe we should move to `try_from` in general, since allocations can always be a point of failure...)

dfalbel commented on crates/ark/src/modules/positron/r_data_explorer.R:
```diff
@@ -346,3 +342,113 @@ write_html <- function(x, include_header) {
+  bin_edges <- h$breaks
+  bin_counts <- h$counts
+
+  # For dates, we convert back the breaks to the date representation.
+  if (inherits(x, "POSIXct")) {
+    bin_edges <- as.POSIXct(h$breaks, tz = attr(x, "tzone"))
+  }
```
Under the hood we'll call `format()` and dependning on the bin edges, it will add the seconds precision. But if bin_edges are round dates, it won't eg:

```
format(as.POSIXct(c("2010-01-01 00:00:00", "2010-01-02 00:00:00")))
```

But it's likely that it will include the seconds precision in mos cases.

dfalbel commented on crates/ark/src/modules/positron/r_data_explorer.R:
```diff
@@ -346,3 +342,113 @@ write_html <- function(x, include_header) {
+      num_bins <- grDevices::nclass.Sturges(x)
+    }
+
+    # The pretty bins algorithm doesn't really make sense for dates,
+    # so we generate our own bin_edges for them.
+    min_value <- min(x)
```
I was unaware of that, from my testing it seemed weird but I didn't go far investigating. I assumed they were simply ignoring pretty for dates.

dfalbel commented on crates/ark/src/modules/positron/r_data_explorer.R:
```diff
@@ -346,3 +342,113 @@ write_html <- function(x, include_header) {
+}
+
+profile_frequency_table <- function(x, limit) {
+    x <- x[!is.na(x)]
```
I think in this case we can show the non-finite values as elements in the freq table

dfalbel commented:
I have addressed most comments. Some pending discussions are: 

- Usage of `as i32` to cast from `i64` into an `RObject`. Since the specific usage in this PR is not problematic I didn't change it, but we can revisit in a future PR.
- Usage of `r_parse_eval0` in tests. Opened https://github.com/posit-dev/positron/issues/4497 to track and will do it once #480 is merged.
